# Install dependencies

In [1]:
!pip install -r requirements.txt

# Import libraries

In [2]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import sqlite3 as sq3

# Data import and cleaning

In [3]:
#import homemade module for SQL interactions
from src.databases import SQLdb

#create a connection to obtain our database and store it in a pandas dataframe.
conn_match = SQLdb().open_connection("data/conduct_match.db")
df_match = pd.read_sql_query("SELECT * FROM match_data", conn_match)
SQLdb().close_connection(conn_match)

In [4]:
#first, look at a snapshot of the data
df_match.head(10)

summonerName         matchId     gameVersion  queueId championName  \
0   Nayumi Shunni  NA1_4815992121  13.21.539.4823      420       Aatrox   
1       Bloalelye  NA1_4815992121  13.21.539.4823      420       LeeSin   
2     ThęNihility  NA1_4815992121  13.21.539.4823      420        Swain   
3         lardcow  NA1_4815992121  13.21.539.4823      420        Vayne   
4             Avô  NA1_4815992121  13.21.539.4823      420    Seraphine   
5  Sakura Useless  NA1_4815992121  13.21.539.4823      420   MonkeyKing   
6       nowin REE  NA1_4815992121  13.21.539.4823      420      Taliyah   
7          T1 Mid  NA1_4815992121  13.21.539.4823      420     Malzahar   
8        platyZ0A  NA1_4815992121  13.21.539.4823      420         Jhin   
9  I never jg lol  NA1_4815992121  13.21.539.4823      420        Leona   

  teamPosition  teamSide  gameDuration  kills  deaths  ...  goldEarned  \
0          TOP       100          1748      9       4  ...       13844   
1       JUNGLE       100          1748      6       3  ...       13652   
2       MIDDLE       100          1748      5       4  ...       11071   
3       BOTTOM       100          1748      4       2  ...       11481   
4      UTILITY       100          1748      1       5  ...        7690   
5          TOP       200          1748      6       5  ...       10826   
6       JUNGLE       200          1748      5       8  ...       10397   
7       MIDDLE       200          1748      2       4  ...        9547   
8       BOTTOM       200          1748      4       2  ...       10930   
9      UTILITY       200          1748      1       6  ...        6568   

   champExperience  visionScore  item0  item1  item2  item3  item5  item6  \
0            16710           18   3161   6692   6694   3111   1037   3340   
1            14674           28   3111   6692   3074   3071   1036   3340   
2            13915           19   1056   3116   3105   3111   1033   3340   
3            12452           42      0   1055   3006   3087   3078   3340   
4            10006           53   3222   3853   3158   2055      0   3364   
5            13387           18   3071   6632   6609   3047      0   3340   
6            11112           33   4645   2421   6655   3158   1052   3364   
7            13362           14   3040   3020   6653   1028   1052   3340   
8            11813           14   3033   3095   6671   3009      0   3363   
9             7637           69   3860   1011   2055   3190   3076   3364   

   victory  
0        1  
1        1  
2        1  
3        1  
4        1  
5        0  
6        0  
7        0  
8        0  
9        0  

[10 rows x 23 columns]

We can see a representation of a single match spread across 10 rows, as evident by the same matchId. Each row represents the stats of an individual player in a game.

In [5]:
df_match.shape

(55280, 23)

With 10 players in each game, we are looking at a total of 5528 games in the database. Since the database schema arranges it so that the summonerName and matchId act as a primary key pair, we can be assured that all matches are unique.

In [6]:
#look at some general details of the data
df_match.describe()

queueId      teamSide  gameDuration         kills        deaths  \
count  55280.0  55280.000000  55280.000000  55280.000000  55280.000000   
mean     420.0    150.000000   1707.634949      5.770658      5.789978   
std        0.0     50.000452    441.734128      4.554377      3.212522   
min      420.0    100.000000    101.000000      0.000000      0.000000   
25%      420.0    100.000000   1489.000000      2.000000      3.000000   
50%      420.0    150.000000   1739.000000      5.000000      6.000000   
75%      420.0    200.000000   1975.000000      8.000000      8.000000   
max      420.0    200.000000   3435.000000     34.000000     31.000000   

            assists    damageDealt    damageTaken    goldEarned  \
count  55280.000000   55280.000000   55280.000000  55280.000000   
mean       7.972938   19295.346690   24852.280644  11220.040720   
std        5.803355   11955.857317   12919.277501   3983.333542   
min        0.000000       0.000000       0.000000    500.000000   
25%        4.000000   10348.750000   15970.500000   8559.000000   
50%        7.000000   17431.000000   23122.000000  11113.000000   
75%       11.000000   25771.000000   31895.000000  13774.250000   
max       47.000000  116818.000000  148952.000000  30388.000000   

       champExperience   visionScore         item0         item1  \
count     55280.000000  55280.000000  55280.000000  55280.000000   
mean      13102.993687     26.563622   3390.708412   3755.199258   
std        4703.881226     21.659908   1863.340509   1838.958572   
min           0.000000      0.000000      0.000000      0.000000   
25%       10092.000000     13.000000   2065.000000   3020.000000   
50%       13110.000000     20.000000   3111.000000   3134.000000   
75%       16154.000000     31.000000   3860.000000   4645.000000   
max       34376.000000    244.000000   8020.000000   8020.000000   

              item2         item3         item5         item6       victory  
count  55280.000000  55280.000000  55280.000000  55280.000000  55280.000000  
mean    3329.679758   3133.470351   1835.953672   3350.767855      0.500000  
std     1713.111687   1741.207893   1794.209623    120.064936      0.500005  
min        0.000000      0.000000      0.000000      0.000000      0.000000  
25%     3006.000000   3006.000000      0.000000   3340.000000      0.000000  
50%     3085.000000   3074.000000   1054.000000   3363.000000      0.500000  
75%     3190.000000   3158.000000   3086.000000   3364.000000      1.000000  
max     8020.000000   8020.000000   8020.000000   3513.000000      1.000000

Let's make sure our data makes sense. 
- Queue ID being 420 for all games is correct since we specifically looked for Ranked Solo/Duo games with that same id number. 
- Team side should have an average of 150 since the "red team" is 100 and the "blue team" is 200, so we expect an equal amount of players from both teams.
- Game duration is in seconds. The shortest game was 101 seconds, which is around the time a team can remake a game if someone fails to connect. The longest game was almost an hour long, which I have personally experienced a handful of times.
- The average player will kill 5.77 other players, die 5.79 times, and get 7.97 assists (which is helping another player get a kill). The reason that deaths are very slightly larger than kills is that a player can die by other means than being killed by another player.
- Damage dealt/taken follows a similar pattern to kills/deaths, in that a player can get damaged in other ways besides other players. They can fight with non-player enemies that deal damage, such as monsters and enemy structures.
- Gold earned seems correct. Every player starts with 500 gold, and since a typical item costs around 3,000 gold to purchase, we expect each player to complete around 3-4 items before the game ends.
- Champion experience points is not something I am very knowledgeable about (it is a number to represent the player's level in the game). From a quick web search, I found that a player starts with 0 experience points and needs a total of 18360 experience points to reach level 18, the max level. The average experience points earned throughout a game is 13103, which is barely over level 15. This sounds logical.
- Vision score is an abstract metric which basically quantifies how many times you helped your team obtain information about the enemy team's location. The characters in the support role (or, in the words of the API, the "UTILITY" role) are typically the ones who can obtain the highest vision score. Since there is typically only one support per team, we see why there is such a big gap between the average of 26.56 vision score versus the highest vision score of 244.
- Each item has a unique identification number. While the mean doesn't have much meaning, one interesting fact we can glean from the data is that the item5 slot is hardly used compared to the others. Each item corresponds to their corresponding key on the keyboard, and the number 5 is far away from other keys used during a game. The reason item6 is used is because people usually change the keybind from 6 to another key.
- Lastly, victory is represented as 1 for a win and 0 for a loss. We expect and find that 0.5 is the average since one team will always win and the other team always loses in a match. Even if a game is remade due to a connection error, the team that voted for a remake is still deemed the "loser" despite none of the players obtaining any rewards or punishments for the game (except the person who failed to connect).

In [7]:
df_match.isnull().sum()

summonerName       0
matchId            0
gameVersion        0
queueId            0
championName       0
teamPosition       0
teamSide           0
gameDuration       0
kills              0
deaths             0
assists            0
damageDealt        0
damageTaken        0
goldEarned         0
champExperience    0
visionScore        0
item0              0
item1              0
item2              0
item3              0
item5              0
item6              0
victory            0
dtype: int64

Thanks to the API and structure of the database, it is not possible for any values to be null in the match database.

In [8]:
df_match.duplicated().sum()

0

We confirm that we have no duplicated rows in our data. Now, we proceed to cleaning the data using the following conditions:
- Remove all rows with gameDuration < 300 seconds (or 5 minutes). This should cover games that are remade due to a connection error, which are not useful data points for our purposes.

In [30]:
df_match.drop(df_match[df_match['gameDuration'] < 300].index, inplace=True)
df_match.shape

(54420, 23)

We went from 5528 games to 5442, meaning we removed 86 games that ended in a remake.

In [9]:
#create a connection to obtain our database and store it in a pandas dataframe.
conn_player = SQLdb().open_connection("data/conduct_player.db")
df_player = pd.read_sql_query("SELECT * FROM player_data", conn_player)
SQLdb().close_connection(conn_player)

In [10]:
df_player.head(10)

summonerName region                                              puuid  \
0   OnlyFullClear    na1  atKGMiKaVsk9i6NPfziLGEzPZISXgD9ruRWk6ESOlu-8rm...   
1     nootherways    na1  G8xJV_OpnMwGUPmKpvtLjqBWabsqQoCDGYJlMqmCH-AUwI...   
2            ߺºß    na1  7zBVk6HEvwUbW6C8VnSvqYxquOPkYM2QkaFwWLk_1v5N6H...   
3       Brimreach    na1  VJfGX2-hgaNckuoFN6e0lT2tuc0nqm-5lxqbGxI0_eLgFH...   
4          ßäłłš     na1  RxF1jmSLJIZYbAmkga8kyEgdtipsIDaoKvn8d-cuUGfcgr...   
5  Savage Prodigy    na1  O-uLtymTa11yknVtgZmqXLelb8iH_jTEtwCTxOP1ESFuih...   
6         JunaLye    na1  xKGg-884LrNHtRcnhK0-IkYCIVWyT_A1sgDCxnLko8ncuK...   
7    FootGooner27    na1  R09G3-h6oxmX6EDAJG_gNus6S8t8OhLxQkBgJrmsLZy0qp...   
8         Vyhaega    na1  82mSnvDzAsa3zD75dNPyHuGbNsJgwUagDmLMcEGddo5Ydv...   
9     HeisenFaker    na1  zITN1XtTfPmWxoNlf7KUSPbul637mvFBmYOaRaaW16TVP-...   

                               leagueId  \
0  5852e915-873e-41b5-bddf-80d78c4e2f9c   
1  d0497be5-58c7-45c6-8c58-1c3cd1d2db38   
2  df960352-50f9-4306-812a-6e827aa69a68   
3  d4caf6a7-b3c7-412c-b05b-ace556b5a728   
4  3d37f5dd-dd80-42f6-b998-fc1bed587c13   
5  c7ea2b2e-db95-4ec3-8eab-7e6a41001603   
6  a2e185b2-df23-4a51-9074-dab2053a31f3   
7  1479063a-fddd-4edd-823c-a3fa8c473151   
8  591385f4-992e-47d6-b57d-f211ff5cf9f3   
9  9b99f93a-99c3-4297-be7a-ca804c120274   

                                          summonerId      tier rank  \
0  nqLB3jST2k5Qlqjh41jOQQo1gAj_oXgboZO19an7JS_5kB...  PLATINUM   II   
1    LKCThiSSoJxP9LgmDZedJKEdckckC13l2gXorol1Cv_tH4o   EMERALD  III   
2    Zq5fbRPHE8mkT27Zbh0zEHIgI915cbJh-vcEqpoku09bAjM   EMERALD  III   
3  qMG5HQKHKvHiZXn0gi-0MmBUSNCO4IgCsM5A_0iMo1kNV6...   EMERALD   IV   
4  yP9IWIyq_-G1j-mZM_vWHEXwwu0KoswpA7d_NpinxIuUGE...  PLATINUM   IV   
5  jFzQ11_7B0sui_zojWDUwN6j7CbWUb-FudWadhcCzFGGhV...   EMERALD  III   
6   AjF9-7j_oiLHdbkcr-oY2T5VFaJyVhI0hpnJ0Nzu5okzqHYP  PLATINUM   IV   
7    SxP0bxj-7n3T83g52B02G_kBiK1DX-DqZD9d5yU-PSu8haM   EMERALD  III   
8  Yo_Co76yiAyp8eMCNJtVC_fiOh0nb2UkSSFOzD-NXoecho...   EMERALD   IV   
9    2MyWR3rahPc77pWVisL0OW42weCety7nD9Bzhnh1tdRbrc8   EMERALD  III   

   leaguePoints  wins  losses  inactive  
0             0    15      25         0  
1            21   247     223         0  
2            31   135     136         0  
3             0   276     272         0  
4            75    13      16         0  
5             0   191     191         0  
6            23    67      75         0  
7            75    34      36         0  
8            50    52      39         0  
9            28    71      82         0

Each row represents the information about a specific player, including their ID numbers and ranked title, as well as their win-loss record.

In [11]:
df_player.shape

(36497, 11)

We should have 36497 unique players in our database. We have 55280 matches, and we expect a roughly concentrated amount of players due to the data scraping method employed (analyzing a handful of matches of a single player then choosing a random match from said player and choosing a random player from said match).

df_player.describe()

- League points (LP) are used in the ranked system for League of Legends. Each tier (Bronze, Silver, Gold, etc) has an associated rank (IV, III, II, I), and you need 100 LP to move up a rank. Starting from Master tier, there are no longer any ranks and their league points become uncapped, which explains the max value of 499 LP.
- A player in our database won 5 more times than lost on average. There could be two reasons for this: 1. a player who plays a lot is more likely to be scraped by our program. A veteran player might be more favored to win than a player who plays less, or 2. players occasionally create new accounts to play in a lower ranked division, which is called "smurfing". These players are able to beat their opponents with ease due to the skill imbalance, creating a dominant win record.
- Inactive is a boolean for inactivity. We see that around 1% of players in this database are inactive. 

# Data analysis

Many players are curious about how champions are performing on average. Let's find the average winrate for a popular champion, Ahri.

In [31]:
ahri_games = df_match[df_match['championName'] == 'Ahri']

In [32]:
ahri_games.head()

summonerName         matchId     gameVersion  queueId championName  \
142     nowin REE  NA1_4807177778  13.20.536.9576      420         Ahri   
182        ßäłłš   NA1_4806418298  13.20.536.9576      420         Ahri   
202  Mochitiger01  NA1_4816166416  13.21.539.4823      420         Ahri   
232  Mochitiger01  NA1_4813923759  13.21.539.4823      420         Ahri   
242  Mochitiger01  NA1_4813804971  13.21.539.4823      420         Ahri   

    teamPosition  teamSide  gameDuration  kills  deaths  ...  goldEarned  \
142       MIDDLE       100          2192      5       6  ...       13868   
182       MIDDLE       100          1979      2       9  ...        9501   
202       MIDDLE       100          1235      7       1  ...        9332   
232       MIDDLE       100          2039      3      10  ...       11374   
242       MIDDLE       100          1818      8       3  ...       11975   

     champExperience  visionScore  item0  item1  item2  item3  item5  item6  \
142            20244           24   1056   1082   6655   3020   4645   3363   
182            12926           22   1056   1033   2031   3020   6655   3340   
202            10462           10   3040   6653   3111   2031      0   3340   
232            16004           26   3040   6653   4629   3111      0   3363   
242            15298           17   3040   3158   6653   4629      0   3340   

     victory  
142        1  
182        0  
202        1  
232        0  
242        1  

[5 rows x 23 columns]

In [33]:
ahri_games.keys()

Index(['summonerName', 'matchId', 'gameVersion', 'queueId', 'championName',
       'teamPosition', 'teamSide', 'gameDuration', 'kills', 'deaths',
       'assists', 'damageDealt', 'damageTaken', 'goldEarned',
       'champExperience', 'visionScore', 'item0', 'item1', 'item2', 'item3',
       'item5', 'item6', 'victory'],
      dtype='object')

In [34]:
ahri_games.shape

(365, 23)

Since every champion in a game is unique, we know that there are 365 games of Ahri in our database.

In [35]:
len(ahri_games[ahri_games['victory'] == 1])/len(ahri_games) * 100

47.671232876712324

Ahri's winrate in our sample is 47.84%. However, though Ahri is usually played in the middle lane of the map (teamPosition = MIDDLE), some people opt to play her in other positions. Let's try to filter out those games and calculate her winrate again.

In [36]:
ahri_mid = ahri_games[ahri_games['teamPosition'] == 'MIDDLE']

In [37]:
ahri_mid.head()

summonerName         matchId     gameVersion  queueId championName  \
142     nowin REE  NA1_4807177778  13.20.536.9576      420         Ahri   
182        ßäłłš   NA1_4806418298  13.20.536.9576      420         Ahri   
202  Mochitiger01  NA1_4816166416  13.21.539.4823      420         Ahri   
232  Mochitiger01  NA1_4813923759  13.21.539.4823      420         Ahri   
242  Mochitiger01  NA1_4813804971  13.21.539.4823      420         Ahri   

    teamPosition  teamSide  gameDuration  kills  deaths  ...  goldEarned  \
142       MIDDLE       100          2192      5       6  ...       13868   
182       MIDDLE       100          1979      2       9  ...        9501   
202       MIDDLE       100          1235      7       1  ...        9332   
232       MIDDLE       100          2039      3      10  ...       11374   
242       MIDDLE       100          1818      8       3  ...       11975   

     champExperience  visionScore  item0  item1  item2  item3  item5  item6  \
142            20244           24   1056   1082   6655   3020   4645   3363   
182            12926           22   1056   1033   2031   3020   6655   3340   
202            10462           10   3040   6653   3111   2031      0   3340   
232            16004           26   3040   6653   4629   3111      0   3363   
242            15298           17   3040   3158   6653   4629      0   3340   

     victory  
142        1  
182        0  
202        1  
232        0  
242        1  

[5 rows x 23 columns]

In [38]:
len(ahri_mid[ahri_mid['victory'] == 1])/len(ahri_mid) * 100

46.80232558139535

Interestingly, Ahri seems to perform worse in her designated role than some other roles. Let's see which champions in particular she struggles the most against. In order to do this, we need to revisit our original dataframe and obtain the enemy lane opponent for each Ahri game.

In [39]:
df_match[df_match['matchId'] == 'NA1_4807177778'][df_match['championName'] != 'Ahri'][df_match['teamPosition'] == 'MIDDLE']['championName'].iloc[0]

/var/folders/5q/_vmdr5x56hb5jwlck0_5vcyr0000gn/T/ipykernel_6450/3925813771.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df_match[df_match['matchId'] == 'NA1_4807177778'][df_match['championName'] != 'Ahri'][df_match['teamPosition'] == 'MIDDLE']['championName'].iloc[0]
/var/folders/5q/_vmdr5x56hb5jwlck0_5vcyr0000gn/T/ipykernel_6450/3925813771.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df_match[df_match['matchId'] == 'NA1_4807177778'][df_match['championName'] != 'Ahri'][df_match['teamPosition'] == 'MIDDLE']['championName'].iloc[0]


'Akali'

In [40]:
ahri_mid.reset_index(drop=True)
enemy_list = []
for id in ahri_mid['matchId']:
    enemy_list.append(df_match[df_match['matchId'] == id][df_match['championName'] != 'Ahri'][df_match['teamPosition'] == 'MIDDLE']['championName'].iloc[0])
ahri_mid['enemyLaner'] = enemy_list

/var/folders/5q/_vmdr5x56hb5jwlck0_5vcyr0000gn/T/ipykernel_6450/1837507391.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  enemy_list.append(df_match[df_match['matchId'] == id][df_match['championName'] != 'Ahri'][df_match['teamPosition'] == 'MIDDLE']['championName'].iloc[0])
/var/folders/5q/_vmdr5x56hb5jwlck0_5vcyr0000gn/T/ipykernel_6450/1837507391.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  enemy_list.append(df_match[df_match['matchId'] == id][df_match['championName'] != 'Ahri'][df_match['teamPosition'] == 'MIDDLE']['championName'].iloc[0])
/var/folders/5q/_vmdr5x56hb5jwlck0_5vcyr0000gn/T/ipykernel_6450/1837507391.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  enemy_list.append(df_match[df_match['matchId'] == id][df_match['championName'] != 'Ahri'][df_match['teamPosition'] == 'MIDDLE']['championName'].iloc[0])
/var/folders/5q/_vmdr5x56hb5jwlck0_5vcyr0000gn/T/ipykernel_6450/

In [100]:
ahri_mid = ahri_mid.reset_index(drop=True)
ahri_mid.head(10)

summonerName         matchId     gameVersion  queueId championName  \
0     nowin REE  NA1_4807177778  13.20.536.9576      420         Ahri   
1        ßäłłš   NA1_4806418298  13.20.536.9576      420         Ahri   
2  Mochitiger01  NA1_4816166416  13.21.539.4823      420         Ahri   
3  Mochitiger01  NA1_4813923759  13.21.539.4823      420         Ahri   
4  Mochitiger01  NA1_4813804971  13.21.539.4823      420         Ahri   
5  Muscular Guy  NA1_4810497099  13.20.536.9576      420         Ahri   
6  OmgItsLegacy  NA1_4815344200  13.21.539.4823      420         Ahri   
7      Corvus P  NA1_4815074537  13.21.539.4823      420         Ahri   
8    MisterWest  NA1_4766823144   13.17.529.525      420         Ahri   
9         hœven  NA1_4760149882  13.17.528.2266      420         Ahri   

  teamPosition  teamSide  gameDuration  kills  deaths  ...  champExperience  \
0       MIDDLE       100          2192      5       6  ...            20244   
1       MIDDLE       100          1979      2       9  ...            12926   
2       MIDDLE       100          1235      7       1  ...            10462   
3       MIDDLE       100          2039      3      10  ...            16004   
4       MIDDLE       100          1818      8       3  ...            15298   
5       MIDDLE       100          1849      4       8  ...            13318   
6       MIDDLE       200          2195      6       8  ...            20020   
7       MIDDLE       100          2364     11       3  ...            22124   
8       MIDDLE       200          1719      5       8  ...            15712   
9       MIDDLE       200          2231      6       7  ...            19723   

   visionScore  item0  item1  item2  item3  item5  item6  victory   enemyLaner  
0           24   1056   1082   6655   3020   4645   3363        1        Akali  
1           22   1056   1033   2031   3020   6655   3340        0       Syndra  
2           10   3040   6653   3111   2031      0   3340        1      Morgana  
3           26   3040   6653   4629   3111      0   3363        0  TwistedFate  
4           17   3040   3158   6653   4629      0   3340        1          Zed  
5           21   4645   6656   2420   3020   3191   3363        0        Diana  
6           30   3157   3158   3135   1058   6653   3363        0       Syndra  
7           31   3020   4645   3089   6653   3108   3364        1     Kassadin  
8           19      0   3157   6656   3102   3020   3363        1     Katarina  
9           25   6656   4628   4645   3157   3020   3363        0       Veigar  

[10 rows x 24 columns]

In [101]:
ahri_mid.shape

(344, 24)

In [111]:
victory_per_champ = ahri_mid.groupby('enemyLaner')['victory'].sum()
games_per_champ = ahri_mid.groupby('enemyLaner')['victory'].count()

victory_per_champ.head(100)

enemyLaner
Akali           5
Akshan          2
Anivia          1
Annie           2
AurelionSol     3
               ..
Yasuo           7
Yone           11
Zed             6
Ziggs           1
Zoe             0
Name: victory, Length: 63, dtype: int64

In [112]:
victory_per_champ.shape

(63,)

In [113]:
games_per_champ.shape

(63,)

In [114]:
winrate_by_champ = victory_per_champ/games_per_champ * 100

In [118]:
winrate_by_champ.head(100)

enemyLaner
Akali          41.666667
Akshan         40.000000
Anivia         50.000000
Annie          33.333333
AurelionSol    42.857143
                 ...    
Yasuo          70.000000
Yone           68.750000
Zed            54.545455
Ziggs          20.000000
Zoe             0.000000
Name: victory, Length: 63, dtype: float64